### PR TITLE
Fix Slack bot handling

### PR DESCRIPTION
**Pull request by lasryaric**
## Description

Fix Slack bot handling of Slack bot users that was introduced in this PR: https://github.com/dust-tt/dust/pull/5048
Everything kept working the same as before this PR, except that the supposedly fix for supporting Zapier was actually broken. Yesterday I was not able to test with Zapier on my local machine, but today I was. Which explains the error in the previous PR.

The key thing to understand in this PR is that Slack has two concepts for bots:
- Bots (apps, workflows)
- And user bots.

I did not find any precise rational for these two types, but basically:
- The Zapier bot is a Slack user bot.
- Workflows are Slack bots.

From our perspective, they are both slack bot, so we treat them the same way, we just load the `SlackUserInfo` type differently. The main change around that is to actually honor the `user.is_bot` flag coming back from Slack, where before we were just looking at the `slackBotId` property from the webhook to define if we were receiving a message from a bot or from a user.

This PR also answers with an appropriate error message when a bot is not whitelisted.

# Screenshots

### Failure case

Screenshot 2024-05-15 at 14 23 04

### Success cases
- Test from Zapier talking to @dust.
Screenshot 2024-05-15 at 14 05 41

- Test from a Slack workflow talking to @dust.
Screenshot 2024-05-15 at 14 05 46
- Test from a user talking to @dust on a public channel.
Screenshot 2024-05-15 at 14 05 49

- Test from a user talking to @Dust in private.
Screenshot 2024-05-15 at 14 36 45

### CODE REVIEW

File: connectors/src/connectors/slack/lib/slack_client.ts [hunk 1]
```diff
@@ -147,7 +128,12 @@ async function getSlackUserInfo(
   }
 
   return {
-    is_bot: false,
+    // Slack has two concept for bots: bots, that you can get through slackClient.bots.info() and
+    // user bots, which are the users related to a bot.
+    // For example, slack workflows are bots, and the Zapier Slack bot is a user bot.
+    // I can't find much more information on why Slack has these two concepts.
+    // From our perspective, a Slack user bot is a bot.
+    is_bot: res.user?.is_bot || false,
```
Comment (flvndvd) [inline connectors/src/connectors/slack/lib/slack_client.ts:136]: `nit`
```suggestion
    // Slack has two concepts for bots:
    // - Bots, that you can get through slackClient.bots.info() and
    // - User bots, which are the users related to a bot.
    // For example, slack workflows are bots, and the Zapier Slack bot is a user bot.
    // Not clear why Slack has these two concepts.
    // From our perspective, a Slack user bot is a bot.
    is_bot: res.user?.is_bot || false,
```

File: connectors/src/connectors/slack/lib/slack_client.ts [hunk 2]
```diff
@@ -117,26 +117,7 @@ export type SlackUserInfo = {
   image_512: string | null;
 };
 
-export async function getSlackUserOrBotInfo(
-  slackClient: WebClient,
-  userId: string
-) {
-  if (userId.startsWith("B")) {
```
Comment (flvndvd) [inline connectors/src/connectors/slack/lib/slack_client.ts:124]: Glad we kill this code 🙃!

File: connectors/src/connectors/slack/lib/workspace_limits.ts [hunk 3]
```diff
@@ -218,15 +219,21 @@ async function isBotAllowed(
...
   if (!whitelist) {
     logger.info(
       { user: slackUserInfo, connectorId: connector.id },
       "Ignoring bot message"
     );
-    return false;
+
+    return new Err(
+      new SlackExternalUserError(
+        "To enable custom interactions between Slack bots and Dust assistants, email us at team@dust.tt"
```
Comment (flvndvd) [inline connectors/src/connectors/slack/lib/workspace_limits.ts:231]: `nit`
```suggestion
        "To enable custom interactions between Slack bots and Dust assistants, email us at team@dust.tt."
```

File: connectors/src/connectors/slack/bot.ts [hunk 4]
```diff
@@ -176,32 +179,37 @@ async function botAnswerMessage(
...
-    {
-      slackChannelId: slackChannel,
-      slackTeamId,
-      slackMessageTs,
-    },
-    slackConfig.whitelistedDomains
-  );
-  if (!hasChatbotAccess) {
-    return new Ok(undefined);
+  if (slackUserInfo.is_bot) {
```
Comment (flvndvd) [inline connectors/src/connectors/slack/bot.ts:193]: Any reason to extract this logic from the main function `notifyIfSlackUserIsNotAllowed` that handles all the permissions logic?

Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts:193]: I think it make sense to separate bot permissions and user permissions.
Also, the code path for user permissions does not follow the existing error model:
We return  a `new Err(SlackExternalUserError('message..'))` everywhere to notify the user, but the function that checks the permissions uses `slack.chat.postMessage()` directly and return undefined. We should fix that in a separate PR.

Comment (flvndvd) [reply connectors/src/connectors/slack/bot.ts:193]: Fair... My point is more that having some of the permission living in `botAnswerMessage` while some other logic lives in `notifyIfSlackUserIsNotAllowed` is not ideal.. While `notifyIfSlackUserIsNotAllowed` might not be the best name for it, the important thing is that the logic is centralized. The more the logic is spread out, the higher the risk of overlooking something.

Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts:193]: Ok, lets deploy that and I will create a task to fix that in a follow up PR.
We have an important customer waiting on it.

Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts:193]: (task created on my personal todo on Slack, will get back to it 💯)

**Review by flvndvd — APPROVED** [review]
Please tackle the comments in a follow up PR 🙏

**Pull request merged**